### PR TITLE
refactor: use `BundleBuilder` as a state builder

### LINKED ISSUE #4614: Integrate BundleBuilder
Integrate bundle builder inside reth: https://github.com/bluealloy/revm/pull/649

comment: https://github.com/paradigmxyz/reth/pull/3512#discussion_r1326447552

### PR DESCRIPTION
**Pull request by tcoratger**
Closes https://github.com/paradigmxyz/reth/issues/4614

Acts as a follow up of https://github.com/paradigmxyz/reth/pull/8264 after the most recent changes.

cc @mattsse

### CODE REVIEW

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -584,39 +592,61 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
...
-                    let present_info = plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1);
-                    entry.insert((present_info, present_info, HashMap::new()))
+            // If the address is not already in the bundle builder's state, add it.
+            if !bundle_builder.get_states().contains(&address) {
+                if let Some(present_info) =
+                    plain_accounts_cursor.seek_exact(address)?.map(|kv| kv.1)
+                {
+                    bundle_builder = bundle_builder
+                        .state_original_account_info(address, present_info.into())
+                        .state_present_account_info(address, present_info.into());
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:638]: looks like we setting functions that take &mut self would be useful on bundlebuilder

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -629,44 +659,49 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
...
+                        .seek_by_key_subkey(*address, storage_key)?
+                        .filter(|s| s.key == storage_key)
                         .is_some()
                     {
                         plain_storage_cursor.delete_current()?
                     }
 
                     // insert value if needed
-                    if *old_storage_value != U256::ZERO {
+                    if storage_slot.original_value() != U256::ZERO {
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:704]:
```suggestion
                    if !storage_slot.original_value().is_zero() {
```

File: crates/storage/provider/src/providers/database/provider.rs [hunk 3]
```diff
@@ -584,39 +592,61 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
         let mut plain_accounts_cursor = self.tx.cursor_write::<tables::PlainAccountState>()?;
         let mut plain_storage_cursor = self.tx.cursor_dup_write::<tables::PlainStorageState>()?;
 
-        let mut reverts: RevertsInit = HashMap::new();
+        // let mut reverts: RevertsInit = HashMap::new();
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:595]:
```suggestion
```

File: crates/storage/provider/src/providers/database/provider.rs [hunk 4]
```diff
@@ -569,13 +573,17 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
...
-        let account_changeset = self.get_or_take::<tables::AccountChangeSets, TAKE>(range)?;
+        let account_changeset =
+            self.get_or_take::<tables::AccountChangeSets, TAKE>(range.clone())?;
 
         // iterate previous value and get plain state value to create changeset
         // Double option around Account represent if Account state is know (first option) and
         // account is removed (Second Option)
 
-        let mut state: BundleStateInit = HashMap::new();
+        // let mut state: BundleStateInit = HashMap::new();
```
Comment (mattsse) [inline crates/storage/provider/src/providers/database/provider.rs:583]:
```suggestion
```

**Review by mattsse — CHANGES_REQUESTED** [review]
ty!
a few cosmetic nits.

this is a bit delicate, so I'd appreciate another review by @rkrasiuk and @rakita

**Comment by quentinv72** [comment]
@tcoratger Just curious why this PR was created? https://github.com/paradigmxyz/reth/pull/8264 is still up-to-date.

**Comment by mattsse** [comment]
@quentinv72 I think this was my bad...

thanks for updating, I'll ensure it is properly reviewed this time...

File: crates/storage/provider/src/providers/database/provider.rs (hunk 1 above)
Comment (tcoratger) [reply crates/storage/provider/src/providers/database/provider.rs:634]: Done here https://github.com/bluealloy/revm/pull/1527

**Comment by mattsse** [comment]
@rkrasiuk idk what to do here

should we proceed with this or #8264

**Comment by tcoratger** [comment]
> @rkrasiuk idk what to do here
> 
> should we proceed with this or #8264

I think you can even simplify some part with the recent updates of revm, including https://github.com/bluealloy/revm/pull/1527

Let me know if we can merge this or not and I have ideas of a follow up so as not to disorganize things here.

**Comment by emhane** [comment]
blocked by https://github.com/paradigmxyz/reth/pull/8264

**Comment by tcoratger** [comment]
> blocked by #8264

This is an equivalent one no @mattsse, following https://github.com/paradigmxyz/reth/pull/8264#issuecomment-2191830400 it seems that #8264 will be staled no?

**Comment by emhane** [comment]
I'd like to merge the other branch first, since you used it as base, so that the commits go to the author. good open source practice is to use `git remote add <repo>` to pull the other persons fork, then use `git cherry-pick` to bring in their changes.

**Pull request merged**